### ISSUE TITLE
Add option for "infinite" scrolling trough Artists, albums, genres ect and next button masked

### ISSUE THREAD

**Issue by antipiot** [deluan/navidrome#132]
Hey!
I would like to be able, if possible, to infinite scroll trough artists, albums, genres ect without having to click on "next" to see more.

Additionnaly, if you select to show like 50 artists, the playbar at the bottom mask the next button. Is this done on purpose? i know you can reduce it but, you know, less click is always better :-)

Thanks once again for your work!

**Comment by deluan** [comment]
I'd love to have it, but I can't find a way to do this with [ReactAdmin](https://marmelab.com/react-admin/), the framework used for the UI. 

Re: player blocking the pagination, I'll work on a fix for it.

**Comment by deluan** [comment]
Maybe we can borrow some ideas from this blog post: https://marmelab.com/blog/2019/01/17/react-timeline.html

**Comment by antipiot** [comment]
I'd like to help you but i have absolutely no knowledge about this :-/

**Comment by deluan** [comment]
@antipiot I pushed a fix for the "player blocking the pagination" issue

**Comment by antipiot** [comment]
> 
> 
> @antipiot I pushed a fix for the "player blocking the pagination" issue

Good job!
Waiting for it to be pushed on the docker :-)

**Comment by deluan** [comment]
I still have work to do on the Album grid view before releasing a new version. Meanwhile if you want, you may use the develop docker image, tag: `deluan/navidrome:develop`

**Comment by antipiot** [comment]
Ho! did not thinked about that! Will do in not time.
Is there a way to send you a beer for your work? Navidrom looks really promising to me, and i've tested everything i could :-)

Airsonic is getting a bit old and navidrom have better handling of albums with more than 1 CD
Funkwhale looks really great but you need 2 login for the Subsonic API
koel looks great, but no subsonic API wich is a real minus

**Comment by certuna** [comment]
I'd love this, but how big is the impact on memory usage, eg with a list of 100,000 songs or 20,000 artists?

**Comment by jvoisin** [comment]
I don't think that anyone has 20,000 artists.

Moreover, the majority of the other subsonic servers are doing it, and they don't make the browser crash.

**Comment by certuna** [comment]
> I don't think that anyone has 20,000 artists.

[sheepishly raises hand] I do...

**Comment by antipiot** [comment]
I'm not used to this, but i guess infinite scrolling will only load what's need to be displayed:
it wont load everything - as a video game dont load every level in memory, it only render what the player is looking at :-)

**Comment by antipiot** [comment]
@deluan , i have an impression this may help:

**Comment by legobuild** [comment]
I realize infinite scrolling might be a ways off, but could you add a higher option for number of items displayed (15/50/100/500)? Also, could that be a setting that gets saved so the user doesn't need to increase it at each logon?

**Comment by deluan** [comment]
We can definitely add more `perPage` options. Will add them for next release

**Comment by esiqveland** [comment]
The `getArtists` call works fine for this in my opionion, you would have to have huge number of artists for that to become unfeasible. However I notice most of the artists in the listing are missing cover art. Would that be much trouble to add?

**Comment by deluan** [comment]
The UI does not use the Subsonic API to browse music, so `getArtists` is not used by it (and this call is actually not paginated). About having too many artists, my library has 1180, which is already a lot for the UI to render, and I know of some users that have as much as 11440 artists! So, not paginating artists is unfeasible until we have infinite scroll.

As for the cover art, Navidrome mimics Subsonic's behaviour: It only get the artist image when there's a call to `getArtistInfo`, which is used to show an artist page in Subsonic clients. Once that call is made, the artist art is cached and served in other calls. This may change in the future, I want to add a worker pool to get this info in background as new artists are added to the library.

**Comment by esiqveland** [comment]
The UI does not need to render all artists at once, only the ones that are currently visible. As long as the response is small enough (in MB) that the client can parse it in a reasonable amount of time, there is no issue to make this at least scrollable. Some mobile clients may experience issues with as many as 11440 artists though, because the payload becomes quite big! Having all the data available also makes it simpler to deal with layouts, as any constraints can be known.

A cached, preloaded artist artwork would be amazing, especially if a a link can be included in the response.

**Comment by elit-altum** [comment]
Hey @deluan !
I would like to take  a shot at this!

**Comment by Vrishabhsk** [comment]
hi i would like to work on this project

**Comment by deluan** [comment]
Adding a reference to a library that can help with this: https://github.com/bvaughn/react-virtualized

**Comment by Balaguru4580** [comment]
I took a look at react-virtualized and it does seem like the library would allow for JIT infinite scroll implementations, but my concern is that the loading needs to be pretty quick to make sure it's not cumbersome when scrolling really fast. This isn't an issue when pagination because it's all already loaded and a user can quickly figure out if what they want isn't on the page.

Also, I think this should be left up as a toggleable option since depending on size, some users may actually just prefer pagination. 

Thanks for linking that library @deluan , by the way! Going to try and implement this on my own and see what happens!

**Comment by Rishabh-malhotraa** [comment]
Suggestion: of adding both the infinite scroll and pagination option

![Screenshot from 2021-04-03 14-48-52](https://user-images.githubusercontent.com/54576074/113474156-cd42bc80-948b-11eb-8b26-c461c5a46606.png)


*Adding an all option in pagination context menu*

**Comment by Balaguru4580** [comment]
We did deliberate over this on Discord and it seems like this is one of those features where it is much more feasible to implement just one of them as the primary system of traversal through lists. 

That being said, I am looking into this, so I appreciate your feedback on this matter @Rishabh-malhotraa

**Comment by pkspyder007** [comment]
@deluan I wanted to work on this issue as my GSOC project. How can I start with it?

**Comment by jvoisin** [comment]
@pkspyder007 everything is documented here: https://www.navidrome.org/docs/gsoc/

**Comment by danshilm** [comment]
Stale bump

**Comment by picklepopper1** [comment]
Is this still something that is still being worked on?

**Comment by deluan** [comment]
Yes, this will be introduced as an option in the upgraded UI, using React-Admin 4: https://marmelab.com/react-admin/InfiniteList.html

**Comment by deluan** [comment]
This will be an option as we move to React-Admin 4, to be released in the upcoming months.

**Comment by tcurdt** [comment]
Is there any update on the new frontend? Just curious. (and unstale)

**Comment by deluan** [comment]
> Is there any update on the new frontend? Just curious. (and unstale)

It is still in progress, ETA ~March/2024, at least the basic stuff. As it won't have all features of the current UI on day one, we plan to make it available in parallel with the current UI, so users can chose which one they want to use. This will continue until we have the new UI on pair with the current one.

**Comment by audiomuze** [comment]
Ok, so today is April Fool's day, there an infinite scroll build yet?

**Comment by patso23** [comment]
I realize this is old but wanted to comment. I just switched to navidrome after using airsonic for years and I really miss the pane on the left that lists all artists. The pagination kills me because sometimes I don't know what I'm looking for and just want to browse through my entire artist list without having to click next a million times

Wish there was a way to greedy load the entire list and cache it for use

**Comment by tcurdt** [comment]
@patso23 I am in ❤️ with [Feishin](https://github.com/jeffvli/feishin) and just skip the navidrome frontend all together these days. Maybe not the worst idea to have the UI separate from the server.

**Comment by deluan** [comment]
I completely agree :)

Originally Navidrome was just a Subsonic server, and the UI was just for admin purposes: Create users, configure transcodings and players. Then I released it to the "public". Soon, due to users demand, I started adding music browsing capabilities to it and then a player.

And it grew a lot since. But I'm not a frontend dev, and the current UI is kind of half-baked.

I plan to rewrite it with a simpler code-base, and even make it available as a stand-alone UI. But this will come after the big refactor the server code is going through right now.

I agree Feishin is great! But you still have to use Navidrome's UI for admin purposes and some features Feishin does not implement (yet), like Shares.

**Comment by add-n2x** [comment]
@deluan Thank you for all your great work!

Re: rewrite of the UI - have you thought about using something like Svelte? I assume it would be performing much better than React, when processing such amounts of data.

**Comment by deluan** [comment]
> Re: rewrite of the UI - have you thought about using something like Svelte? I assume it would be performing much better than React, when processing such amounts of data.

I don't think I'll have time to learn Svelte just to work on Navidrome. I use whatever I already use in my dayjob or I know how to code on (or something interested enough to prompt me to learn it :)

Anyway, what makes it slow is not React, it is how React-Admin renders lists/grids. It needs to do some virtualization for large lists. I plan to do that in the new UI

**Comment by Lenzelot** [comment]
@deluan 2-years navidrome users here. Awesome software! 

One thing consistently bugs me is that on bigger playlists I constantly have to change items per page to 500. Instead of implementing the whole infinite scrolling thing, which sure isn't that easy to do, could you just give us an option to change default items per page in the web player user settings? In my experience the web UI seems to handle so many entries well. I think it would solve the problem for most of us and would be simpler to implement. Anyway, your decision what to do of course.

Thanks for your contributions and considerations.

**Comment by ghost** [comment]
Agreed on 500 selection. That would solve a lot!! Also to make it a permanent selection
